subsys/jwt/jwt.c

```diff
@@ -317,3 +395,125 @@ int jwt_init_builder(struct jwt_builder *builder,
 
 	return jwt_add_header(builder);
 }
+
+int jwt_parse_payload(struct jwt_parser *parser, int32_t *expt, int32_t *iat, char *aud)
+{
+	struct jwt_payload payload;
+	int res;
+
+	res = base64url_decode(parser->buf, parser->buf_len, parser->payload, parser->payload_len);
+	if (res < 0) {
+		return res;
+	}
+
+	res = json_obj_parse(parser->buf, res, jwt_payload_desc, ARRAY_SIZE(jwt_payload_desc),
+			     &payload);
+	if (res == (1 << ARRAY_SIZE(jwt_payload_desc)) - 1) {
+		*expt = payload.exp;
+		*iat = payload.iat;
+		strcpy(aud, payload.aud);
+		res = 0;
+	} else if (res >= 0) {
+		res = -EINVAL;
+	}
+
+	return res;
+}
+
+int jwt_verify(struct jwt_parser *parser, const char *der_key, size_t der_key_len)
+{
+	struct jwt_builder builder;
+	char *builder_header;
+	size_t builder_header_len;
+	char *builder_sign;
+	size_t builder_sign_len;
+	int res;
+
+	/*
+	 * Use existing jwt_add_header and jwt_sign functions to fill a
+	 * jwt_builder structure and use it to compare header and signature
+	 * fields
+	 */
+
+	/*
+	 * Check JWT header
+	 */
+
+	res = jwt_init_builder(&builder, parser->buf, parser->buf_len);
+	if (res != 0) {
+		return res;
+	}
+	builder_header = builder.base;
+	builder_header_len = builder.buf - builder.base;
+	res = strncmp(parser->header, builder_header, builder_header_len);
```

Comment:
This is not correct, the strncmp will only check that the `builder_header` matches a prefix of `parser->header`. Is there a concern that the builder somehow didn't get a terminating nul?

Frankly, in this case, just use strcmp, since you are controlling the builder, and the buffer should always have the null termination.

Reply:
> Is there a concern that the builder somehow didn't get a terminating nul?

Yes, there is.

`jwt_init_parser` uses `strchr` to find payload and signature positions in the token but `.` (dot separators) are not overwrittent by `\0` to avoid modifying the original `jwt` input string.

So in `tests/subsys/jwt/src/main.c` example:

```C
jwt = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJpb3Qtd29yay0xOTk0MTkiLCJleHAiOjE1MzAzMTIwMjYsImlhdCI6MTUzMDMwODQyNn0.VHKDT-dvWW3UDBGokkttP7P6DNDnd4jjz_wLsOvn9RVBoDAAVG1mQnlROPtOHyRTotWzFYf2g02-VyZoh326yDFXM-66U7ycqyUsJdWBZWYqub6iZ17l0hjodoSpZBJqPB6F9c8Q5u9x1gWVw2l2XNASQyTDIZ3db_GiQbx1SYnNtyxsospoKClIebee-EQkcut9jBPNyiKrzZRPuX0xCiPJYTumHdHSp0UPIDIij2hJl_tacXs-hz_ckqT5HZvJPsxgJCeXz3tP7IlgbeGbPFMFmaYNX4UxYD4XS10MPXUK4_qp8wK4EqjT2YBNF7xXi_sjWMj7isc0i0YSIl6_nQ"

parser->header = "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJpb3Qtd29yay0xOTk0MTkiLCJleHAiOjE1MzAzMTIwMjYsImlhdCI6MTUzMDMwODQyNn0.VHKDT-dvWW3UDBGokkttP7P6DNDnd4jjz_wLsOvn9RVBoDAAVG1mQnlROPtOHyRTotWzFYf2g02-VyZoh326yDFXM-66U7ycqyUsJdWBZWYqub6iZ17l0hjodoSpZBJqPB6F9c8Q5u9x1gWVw2l2XNASQyTDIZ3db_GiQbx1SYnNtyxsospoKClIebee-EQkcut9jBPNyiKrzZRPuX0xCiPJYTumHdHSp0UPIDIij2hJl_tacXs-hz_ckqT5HZvJPsxgJCeXz3tP7IlgbeGbPFMFmaYNX4UxYD4XS10MPXUK4_qp8wK4EqjT2YBNF7xXi_sjWMj7isc0i0YSIl6_nQ"
parser->header_len = 36

parser->payload = "eyJhdWQiOiJpb3Qtd29yay0xOTk0MTkiLCJleHAiOjE1MzAzMTIwMjYsImlhdCI6MTUzMDMwODQyNn0.VHKDT-dvWW3UDBGokkttP7P6DNDnd4jjz_wLsOvn9RVBoDAAVG1mQnlROPtOHyRTotWzFYf2g02-VyZoh326yDFXM-66U7ycqyUsJdWBZWYqub6iZ17l0hjodoSpZBJqPB6F9c8Q5u9x1gWVw2l2XNASQyTDIZ3db_GiQbx1SYnNtyxsospoKClIebee-EQkcut9jBPNyiKrzZRPuX0xCiPJYTumHdHSp0UPIDIij2hJl_tacXs-hz_ckqT5HZvJPsxgJCeXz3tP7IlgbeGbPFMFmaYNX4UxYD4XS10MPXUK4_qp8wK4EqjT2YBNF7xXi_sjWMj7isc0i0YSIl6_nQ"
parser->payload_len = 79

parser->sign = "VHKDT-dvWW3UDBGokkttP7P6DNDnd4jjz_wLsOvn9RVBoDAAVG1mQnlROPtOHyRTotWzFYf2g02-VyZoh326yDFXM-66U7ycqyUsJdWBZWYqub6iZ17l0hjodoSpZBJqPB6F9c8Q5u9x1gWVw2l2XNASQyTDIZ3db_GiQbx1SYnNtyxsospoKClIebee-EQkcut9jBPNyiKrzZRPuX0xCiPJYTumHdHSp0UPIDIij2hJl_tacXs-hz_ckqT5HZvJPsxgJCeXz3tP7IlgbeGbPFMFmaYNX4UxYD4XS10MPXUK4_qp8wK4EqjT2YBNF7xXi_sjWMj7isc0i0YSIl6_nQ"
parser->sign_len = 342
```

Reply:
Two possibilities. Change the dot to a `\0` and then put it back. Or do something complicated that does the comparison. If you know where the dot was you could check the length manually, and if equal, just do a memcmp.

Reply:
> That may be the case, but the strncmp is still incorrect, as it only compares the prefix. e.g. `strncmp("Hello", "HelloWorld", 5) == 0`.

Understood.

> If you know where the dot was you could check the length manually, and if equal, just do a memcmp.

I added this check.